File: 07_linked_list/07_02_reverse_linked_list/reverse_linked_list.cpp
#include <iostream>

class ListNode
{
public:
    int val;
    ListNode *next;

    ListNode(int value = 0, ListNode *nextNode = nullptr) : val(value), next(nextNode) {}
};

ListNode *reverseList(ListNode *head)
{
    ListNode *prev = nullptr;
    ListNode *current = head;
    while (current != nullptr)
    {
        ListNode *nextNode = current->next;
        current->next = prev;
        prev = current;
        current = nextNode;
    }
    return prev;
}

int main()
{
    // Example usage
    ListNode *head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    std::cout << "Before reverse: ";
    ListNode *trav = head;
    while (trav != nullptr)
    {
        std::cout << trav->val << " ";
        trav = trav->next;
    }
    std::cout << std::endl;

    ListNode *reversedHead = reverseList(head);

    // Print the reversed list for verification
    std::cout << "After reverse: ";
    ListNode *reversedTrav = reversedHead;
    while (reversedTrav != nullptr)
    {
        std::cout << reversedTrav->val << " ";
        reversedTrav = reversedTrav->next;
    }
    std::cout << std::endl;

    // Remember to free allocated memory
    delete head;
    delete reversedHead;

    return 0;
}

File: 07_linked_list/07_02_reverse_linked_list/reverse_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverse_list(head):
    prev = None
    current = head
    while current is not None:
        next_node = current.next
        current.next = prev
        prev = current
        current = next_node
    return prev

# Example usage
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
print("Before reverse:", end=" ")
trav = head
while trav is not None:
    print(trav.val, end=" ")
    trav = trav.next
print()

reversed_head = reverse_list(head)

# Print the reversed list for verification
print("After reverse:", end=" ")
while reversed_head is not None:
    print(reversed_head.val, end=" ")
    reversed_head = reversed_head.next
